src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Lookup.txt

```diff
@@ -0,0 +1,70 @@
+﻿// Basic LookUp
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 2)
+{Id:2,Name:"Element 2"}
+
+// Not found
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 3)
+Blank()
+
+// Basic LookUp with reduction
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 2, Name)
+"Element 2"
+
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 3, Name)
+Blank()
+
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 2).Name
+"Element 2"
+
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2"}), Id = 3).Name
+Blank()
+
+// Return first result only
+>> LookUp(Table({ Id: 1, Name: "Element 1" }, { Id:2, Name: "Element 2a"}, { Id:2, Name: "Element 2b"}), Id = 2).Name
+"Element 2a"
+
+// 'As' with LookUp
+>> LookUp([1,2,3] As X, X.Value > 2)
+{Value:3}
+
+>> LookUp([1,2,3] As X, X.Value > 2).Value
+3
+
+>> LookUp([1,2,3] As X, X.Value > 2, X.Value)
+3
+
+// 'ThisRecord is implicit'
+>> LookUp([1,2,3], Value >= 2)
+{Value:2}
+
+>> LookUp([1,2,3], Value >= 2).Value
+2
+
+>> LookUp([1,2,3], Value >= 2, Value)
+2
+
+>> LookUp([1,2,3], ThisRecord.Value >= 2)
+{Value:2}
+
+>> LookUp([1,2,3], ThisRecord.Value >= 2, Value)
+2
+
+>> LookUp([1,2,3], ThisRecord.Value > 2).Value
+3
+
+// $$$ Corner cases like  negative indices, N > length, doubles instead of int, empty. 
+
+>> LookUp([1,2,3], ThisRecord.Value > 3)
```

Comment:
More interesting tests:
- Lookups in very long lists: `LookUp(Sequence(30000), Value > 25000, Value * 2) ==> 50002` or `LookUp(Sequence(30000), Value > 35000, Value * 2) ==> Blank()`

Reply:
Re long lists - 30000 is a big number ... would 300 suffice? I don't want to just burn cpu cycles.

Reply:
We've had bugs in the past that start showing up with tables around 10000 elements... I would hope that we could deal with ~30k+ items in memory fairly quickly

Reply:
@CarlosFigueiraMSFT these are interesting tests, but they are not specific to the lookup function? They are tests to check if errors fall through correctly, which the implementation of the function doesn't deal with directly. The ones returning an error at runtime can be properly caught (like Div0), but those that throw an exception during the formula check (for example, where the types are incompatible like Len("string" > 3) cannot be "caught" with IfError or checking the tests result for Error (with or without a specific Kind).
Happy to take your examples (and ensure to use runtime errors we can check for) but they aren't tests for LookUp code specifically? Let me know what you'd like me to do.

Reply:
Correction - it does return the errorvalue so that should be checked, but any one error is sufficient as a test for lookup()

Reply:
@jorisdg I see those tests not only as "tests" per se (i.e., preventing regressions), but also as a de facto specification for the function. Also, in this C# implementation we have a wrapper that deals with errors, but it's possible that in another language / platform they won't have such a thing, so having those tests that clearly define what the behavior of the function wrt errors is useful to help the language be consistent across different platforms.

Reply:
The long list tests seem to average between 250 and 300 ms each when running on my machine (Xeon with 3.6Ghz cores).

Reply:
Invalid argument tests for each argument would be good as well.
`LookUp("123", true)`
`LookUp(["test"], 123)`

Reply:
@emhomm4 I would agree, but those throw exceptions and don't return Error() values so I don't see a way to define a test like that?

Reply:
Exceptions? Or compile time errors?
For compile time errors, see the last few tests in 
https://github.com/microsoft/Power-Fx/blob/9938c06b7d2c4d01732d5f535d4314fbb0f98d11/src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt

Reply:
If it's exceptions, than that's concerning, and might need emily's #210 to go in first

Reply:
Compile time errors. Thanks for the PR link, I will try that out!!

Reply:
Errors added.